### ISSUE TITLE
`np.stack` does not respect stickied arrays

### ISSUE THREAD

**Issue by romanngg** [google/jax#2226]
In the snipped below, I stick `a` to the CPU. Addition and concatenation respects this, but `np.stack`, for some reason, does not:

```python

import jax

a = jax.random.normal(jax.random.PRNGKey(2), (10, 10))
b = jax.random.normal(jax.random.PRNGKey(2), (10, 10))

a.device_buffer.device(), b.device_buffer.device()
(GpuDevice(id=0), GpuDevice(id=0))

a = jax.device_put(a, jax.devices('cpu')[0])
a.device_buffer.device()
CpuDevice(id=0)

jax.np.stack([a, b]).device_buffer.device()
GpuDevice(id=0)  # Must be CPU!

(a + b).device_buffer.device()
CpuDevice(id=0)

jax.np.concatenate([a, b]).device_buffer.device()
CpuDevice(id=0)
```

**Comment by romanngg** [comment]
Friendly ping :) I'm wondering if this and related issue https://github.com/google/jax/issues/2225 are feasible to fix in the near term, or perhaps these aren't simple bugs and need a lot of change?

This would help us remove a lot of device-managing complexity in neural-tangents and allow some useful unsupported currently behavior. Thanks a lot!

**Comment by madisonmay** [comment]
Started digging into this tonight.

First issue seems to be that `jax.numpy.array()` calls `device_put` on any `jax.numpy.ndarray` or `jax.numpy.ndarray` object it receives:

```
@_wraps(onp.array)
def array(object, dtype=None, copy=True, order="K", ndmin=0):
  ...
  if isinstance(object, ndarray):
    if dtype and _dtype(object) != dtypes.canonicalize_dtype(dtype):
      out = lax.convert_element_type(object, dtype)
    else:
      out = device_put(object)
```

Seems like this could maybe be fixed with a secondary isinstance check?

```
@_wraps(onp.array)
def array(object, dtype=None, copy=True, order="K", ndmin=0):
  ...
  if isinstance(object, ndarray):
    if dtype and _dtype(object) != dtypes.canonicalize_dtype(dtype):
      out = lax.convert_element_type(object, dtype)
    elif isinstance(object, DeviceArray):
      # respect current device assignment
      out = object
    else:
      out = device_put(object)
```

This causes `hstack` and `vstack` to pin to CPU (whereas they previously did not).

However, `stack` still results in GPU placement.  I traced that down to `lax.concatenate` -- in particular it's behavior when it's inputs are reshaped such that one axis is size 1. Seems to be irrespective of what axis is passed.

```
import jax
import jax.numpy as np

a = jax.random.normal(jax.random.PRNGKey(2), (10, 10))
b = jax.random.normal(jax.random.PRNGKey(3), (10, 10))

a = jax.device_put(a, jax.devices('cpu')[0])

# prints cpu:0
print(jax.lax.concatenate([a, b], 0).device_buffer.device()) 

# prints gpu:0
print(jax.lax.concatenate([np.expand_dims(a, 0), np.expand_dims(b, 0)], 0).device_buffer.device()) 

# prints gpu:0
print(jax.lax.concatenate([np.expand_dims(a, 0), np.expand_dims(b, 0)], 1).device_buffer.device()) 

# prints cpu:0
print(jax.lax.concatenate([np.reshape(a, [2, 5, 10]), np.reshape(b, [2, 5, 10])], 0).device_buffer.device()) 
```

So... broadcasting and reshape related. Dug into the lax.reshape operation and found this:

```
# We have a nonstandard reshape impl so that we can be lazy about data movement.
def _reshape_impl(operand, new_sizes, dimensions):
  old_sizes = onp.shape(operand)
  if type(operand) is xla.DeviceArray and dimensions is None:
    bcast_dims = _is_singleton_reshape(old_sizes, new_sizes)
    if bcast_dims is not None:
      aval = ShapedArray(new_sizes, operand.dtype)
      lazy_expr = lazy.broadcast(operand._lazy_expr, new_sizes, bcast_dims)
      return xla.DeviceArray(aval, None, lazy_expr, operand.device_buffer)
```

There's a new `xla.DeviceArray` created which is passed None for the device.  This means that even when all inputs to a given operation are on cpu we get a GPU result.

```
# prints gpu:0
print(jax.lax.concatenate([np.expand_dims(a, 0), np.expand_dims(a, 0)], 1).device_buffer.device()) 
```

I'm not sure I can just pass `operand._device` here / I'm sure there's a good reason this is done?  But hopefully this is close enough for someone who actually knows their way around the codebase to get this over the finish line.

**Comment by gnecula** [comment]
Thank you! Your proposed bug fix is correct. This will be covered by #2882.
